Mathlib/Algebra/Lie/Rank.lean

```diff
@@ -0,0 +1,251 @@
+/-
+Copyright (c) 2024 Johan Commelin. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Johan Commelin
+-/
+import Mathlib.Algebra.Lie.OfAssociative
+import Mathlib.Data.MvPolynomial.Monad
+import Mathlib.LinearAlgebra.Charpoly.ToMatrix
+import Mathlib.LinearAlgebra.Dimension.Finrank
+import Mathlib.LinearAlgebra.FreeModule.StrongRankCondition
+import Mathlib.LinearAlgebra.Matrix.Charpoly.Univ
+
+/-!
+# Rank of a Lie algebra and regular elements
+
+Let `L` be a Lie algebra over a nontrivial commutative ring `R`,
+and assume that `L` is finite free as `R`-module.
+Then the coefficients of the characteristic polynomial of `ad R L x` are polynomial in `x`.
+The *rank* of `L` is the smallest `n` for which the `n`-th coefficient is not the zero polynomial.
+
+Continuing to write `n` for the rank of `L`, an element `x` of `L` is *regular*
+if the `n`-th coefficient of the characteristic polynomial of `ad R L x` is non-zero.
+
+## Main declarations
+
+* `LieAlgebra.rank R L` is the rank of a Lie algebra `L` over a commutative ring `R`.
+* `LieAlgebra.IsRegular R x` is the predicate that an element `x` of a Lie algebra `L` is regular.
+
+## References
+
+* [barnes1967]: "On Cartan subalgebras of Lie algebras" by D.W. Barnes.
+
+-/
+
+open scoped BigOperators
+
+variable {R L M ι ι' ιM ιM' : Type*}
```

Comment:
None of the primed variables are used. I expect you have plans to use them but I think we should only add them when that happens.
```suggestion
variable {R L M ι ιM : Type*}
```